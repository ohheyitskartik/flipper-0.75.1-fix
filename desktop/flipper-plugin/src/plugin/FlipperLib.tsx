/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {Logger} from '../utils/Logger';
import {RealFlipperDevice} from './DevicePlugin';
import {NormalizedMenuEntry} from './MenuEntry';
import {RealFlipperClient} from './Plugin';

/**
 * This interface exposes all global methods for which an implementation will be provided by Flipper itself
 */
export interface FlipperLib {
  logger: Logger;
  enableMenuEntries(menuEntries: NormalizedMenuEntry[]): void;
  createPaste(input: string): Promise<string | undefined>;
  GK(gatekeeper: string): boolean;
  isPluginAvailable(
    device: RealFlipperDevice,
    client: RealFlipperClient | null,
    pluginId: string,
  ): boolean;
  selectPlugin(
    device: RealFlipperDevice,
    client: RealFlipperClient | null,
    pluginId: string,
    deeplink: unknown,
  ): void;
}
