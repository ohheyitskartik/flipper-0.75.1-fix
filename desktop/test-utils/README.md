# flipper-test-utils

`flipper-test-utils` is a library containing common Flipper test utilities.
