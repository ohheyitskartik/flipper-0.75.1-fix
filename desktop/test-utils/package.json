{
  "name": "flipper-test-utils",
  "version": "0.0.0",
  "description": "Library containing common Flipper test utilities",
  "repository": "facebook/flipper",
  "main": "lib/index.js",
  "flipperBundlerEntry": "src",
  "types": "lib/index.d.ts",
  "license": "MIT",
  "bugs": "https://github.com/facebook/flipper/issues",
  "scripts": {
    "reset": "rimraf lib *.tsbuildinfo",
    "build": "tsc -b",
    "prepack": "yarn reset && yarn build"
  },
  "files": [
    "lib/**/*"
  ],
  "homepage": "https://github.com/facebook/flipper",
  "keywords": [
    "Flipper"
  ],
  "author": "Facebook, Inc"
}
