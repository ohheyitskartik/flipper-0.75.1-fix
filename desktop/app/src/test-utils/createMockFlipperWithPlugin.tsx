/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import React from 'react';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {
  render,
  RenderResult,
  act as testingLibAct,
} from '@testing-library/react';
import {queries} from '@testing-library/dom';

import {
  selectPlugin,
  starPlugin,
  selectDevice,
  selectClient,
} from '../reducers/connections';
import BaseDevice from '../devices/BaseDevice';

import {rootReducer} from '../store';
import {Store} from '../reducers/index';
import Client, {ClientQuery} from '../Client';

import {buildClientId} from '../utils/clientUtils';
import {Logger} from '../fb-interfaces/Logger';
import {PluginDefinition} from '../plugin';
import {registerPlugins} from '../reducers/plugins';
import PluginContainer from '../PluginContainer';
import {getPluginKey, isDevicePluginDefinition} from '../utils/pluginUtils';
import {getInstance} from '../fb-stubs/Logger';
import {initializeFlipperLibImplementation} from '../utils/flipperLibImplementation';

export type MockFlipperResult = {
  client: Client;
  device: BaseDevice;
  store: Store;
  pluginKey: string;
  sendMessage(method: string, params: any, client?: Client): void;
  createDevice(serial: string): BaseDevice;
  createClient(
    device: BaseDevice,
    name: string,
    query?: ClientQuery,
    skipRegister?: boolean,
  ): Promise<Client>;
  logger: Logger;
  togglePlugin(plugin?: string): void;
};

type MockOptions = Partial<{
  /**
   * can be used to intercept outgoing calls. If it returns undefined
   * the base implementation will be used
   */
  onSend?: (pluginId: string, method: string, params?: object) => any;
  additionalPlugins?: PluginDefinition[];
  dontEnableAdditionalPlugins?: true;
  asBackgroundPlugin?: true;
}>;

export async function createMockFlipperWithPlugin(
  pluginClazz: PluginDefinition,
  options?: MockOptions,
): Promise<MockFlipperResult> {
  const store = createStore(rootReducer);
  const logger = getInstance();
  initializeFlipperLibImplementation(store, logger);
  store.dispatch(
    registerPlugins([pluginClazz, ...(options?.additionalPlugins ?? [])]),
  );

  function createDevice(serial: string): BaseDevice {
    const device = new BaseDevice(
      serial,
      'physical',
      'MockAndroidDevice',
      'Android',
    );
    store.dispatch({
      type: 'REGISTER_DEVICE',
      payload: device,
    });
    device.loadDevicePlugins(store.getState().plugins.devicePlugins);
    return device;
  }

  async function createClient(
    device: BaseDevice,
    name: string,
    query?: ClientQuery,
    skipRegister?: boolean,
  ): Promise<Client> {
    query = query ?? {
      app: name,
      os: 'Android',
      device: device.title,
      device_id: device.serial,
      sdk_version: 4,
    };
    const id = buildClientId({
      app: query.app,
      os: query.os,
      device: query.device,
      device_id: query.device_id,
    });

    const client = new Client(
      id,
      query,
      null, // create a stub connection to avoid this plugin to be archived?
      logger,
      store,
      [
        ...(isDevicePluginDefinition(pluginClazz) ? [] : [pluginClazz.id]),
        ...(options?.dontEnableAdditionalPlugins
          ? []
          : options?.additionalPlugins?.map((p) => p.id) ?? []),
      ],
      device,
    );

    client.rawCall = async (
      method: string,
      _fromPlugin: boolean,
      params: any,
    ): Promise<any> => {
      const intercepted = options?.onSend?.(method, params);
      if (intercepted !== undefined) {
        return intercepted;
      }
      switch (method) {
        case 'getPlugins':
          // assuming this plugin supports all plugins for now
          return {
            plugins: [...store.getState().plugins.clientPlugins.keys()],
          };
        case 'getBackgroundPlugins':
          return {plugins: options?.asBackgroundPlugin ? [pluginClazz.id] : []};
        default:
          throw new Error(
            `Test client doesn't support rawCall method '${method}'`,
          );
      }
    };
    client.rawSend = jest.fn();

    // enable the plugin
    if (
      !isDevicePluginDefinition(pluginClazz) &&
      !store
        .getState()
        .connections.userStarredPlugins[client.query.app]?.includes(
          pluginClazz.id,
        )
    ) {
      store.dispatch(
        starPlugin({
          plugin: pluginClazz,
          selectedApp: client.query.app,
        }),
      );
    }
    if (!options?.dontEnableAdditionalPlugins) {
      options?.additionalPlugins?.forEach((plugin) => {
        if (!isDevicePluginDefinition(plugin)) {
          store.dispatch(
            starPlugin({
              plugin,
              selectedApp: client.query.app,
            }),
          );
        }
      });
    }
    await client.init();

    // As convenience, by default we select the new client, star the plugin, and select it
    if (!skipRegister) {
      store.dispatch({
        type: 'NEW_CLIENT',
        payload: client,
      });
    }
    return client;
  }

  const device = createDevice('serial');
  const client = await createClient(device, 'TestApp');

  store.dispatch(selectDevice(device));
  store.dispatch(selectClient(client.id));

  store.dispatch(
    selectPlugin({
      selectedPlugin: pluginClazz.id,
      selectedApp: client.query.app,
      deepLinkPayload: null,
      selectedDevice: device,
    }),
  );

  return {
    client,
    device: device as any,
    store,
    sendMessage(method, params, actualClient = client) {
      actualClient.onMessage(
        JSON.stringify({
          method: 'execute',
          params: {
            api: pluginClazz.id,
            method,
            params,
          },
        }),
      );
    },
    createDevice,
    createClient,
    logger,
    pluginKey: getPluginKey(client.id, device, pluginClazz.id),
    togglePlugin(id?: string) {
      const plugin = id
        ? store.getState().plugins.clientPlugins.get(id)
        : pluginClazz;
      if (!plugin) {
        throw new Error('unknown plugin ' + id);
      }
      store.dispatch(
        starPlugin({
          plugin,
          selectedApp: client.query.app,
        }),
      );
    },
  };
}

type Renderer = RenderResult<typeof queries>;

export async function renderMockFlipperWithPlugin(
  pluginClazz: PluginDefinition,
  options?: MockOptions,
): Promise<
  MockFlipperResult & {
    renderer: Renderer;
    act: (cb: () => void) => void;
  }
> {
  const args = await createMockFlipperWithPlugin(pluginClazz, options);

  function selectTestPlugin(store: Store, client: Client) {
    store.dispatch(
      selectPlugin(
        isDevicePluginDefinition(pluginClazz)
          ? {
              selectedPlugin: pluginClazz.id,
              selectedApp: null,
              deepLinkPayload: null,
              selectedDevice: store.getState().connections.selectedDevice!,
            }
          : {
              selectedPlugin: pluginClazz.id,
              selectedApp: client.query.app,
              deepLinkPayload: null,
              selectedDevice: store.getState().connections.selectedDevice!,
            },
      ),
    );
  }

  selectTestPlugin(args.store, args.client);

  const renderer = render(
    <Provider store={args.store}>
      <PluginContainer logger={args.logger} />
    </Provider>,
  );

  return {
    ...args,
    renderer: renderer as any,
    act(cb) {
      testingLibAct(cb);
      args.client.flushMessageBuffer();
    },
  };
}
