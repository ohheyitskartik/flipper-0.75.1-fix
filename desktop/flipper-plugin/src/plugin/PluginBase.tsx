/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {SandyPluginDefinition} from './SandyPluginDefinition';
import {EventEmitter} from 'events';
import {Atom} from '../state/atom';
import {MenuEntry, NormalizedMenuEntry, normalizeMenuEntry} from './MenuEntry';
import {FlipperLib} from './FlipperLib';
import {Device, RealFlipperDevice} from './DevicePlugin';
import {batched} from '../state/batch';
import {Idler} from '../utils/Idler';
import {message} from 'antd';

type StateExportHandler<T = any> = (
  idler: Idler,
  onStatusMessage: (msg: string) => void,
) => Promise<T>;
type StateImportHandler<T = any> = (data: T) => void;

export interface BasePluginClient {
  readonly device: Device;

  /**
   * the onDestroy event is fired whenever a device is unloaded from Flipper, or a plugin is disabled.
   */
  onDestroy(cb: () => void): void;

  /**
   * the onActivate event is fired whenever the plugin is actived in the UI
   */
  onActivate(cb: () => void): void;

  /**
   * The counterpart of the `onActivate` handler.
   */
  onDeactivate(cb: () => void): void;

  /**
   * Triggered when this plugin is opened through a deeplink
   */
  onDeepLink(cb: (deepLink: unknown) => void): void;

  /**
   * Triggered when the current plugin is being exported and should create a snapshot of the state exported.
   * Overrides the default export behavior and ignores any 'persist' flags of state.
   */
  onExport<T = any>(exporter: StateExportHandler<T>): void;

  /**
   * Triggered directly after the plugin instance was created, if the plugin is being restored from a snapshot.
   * Should be the inverse of the onExport handler
   */
  onImport<T = any>(handler: StateImportHandler<T>): void;

  /**
   * Register menu entries in the Flipper toolbar
   */
  addMenuEntry(...entry: MenuEntry[]): void;

  /**
   * Creates a Paste (similar to a Github Gist).
   * Facebook only function. Resolves to undefined if creating a paste failed.
   */
  createPaste(input: string): Promise<string | undefined>;

  /**
   * Returns true if the user is taking part in the given gatekeeper.
   * Always returns `false` in open source.
   */
  GK(gkName: string): boolean;
}

let currentPluginInstance: BasePluginInstance | undefined = undefined;

export function setCurrentPluginInstance(
  instance: typeof currentPluginInstance,
) {
  currentPluginInstance = instance;
}

export function getCurrentPluginInstance(): typeof currentPluginInstance {
  return currentPluginInstance;
}

export abstract class BasePluginInstance {
  /** generally available Flipper APIs */
  readonly flipperLib: FlipperLib;
  /** the original plugin definition */
  definition: SandyPluginDefinition;
  /** the plugin instance api as used inside components and such  */
  instanceApi: any;
  /** the device owning this plugin */
  readonly device: Device;

  activated = false;
  destroyed = false;
  readonly events = new EventEmitter();

  // temporarily field that is used during deserialization
  initialStates?: Record<string, any>;

  // all the atoms that should be serialized when making an export / import
  readonly rootStates: Record<string, Atom<any>> = {};
  // last seen deeplink
  lastDeeplink?: any;
  // export handler
  exportHandler?: StateExportHandler;
  // import handler
  importHandler?: StateImportHandler;

  menuEntries: NormalizedMenuEntry[] = [];

  constructor(
    flipperLib: FlipperLib,
    definition: SandyPluginDefinition,
    realDevice: RealFlipperDevice,
    initialStates?: Record<string, any>,
  ) {
    this.flipperLib = flipperLib;
    this.definition = definition;
    this.initialStates = initialStates;
    if (!realDevice) {
      throw new Error('Illegal State: Device has not yet been loaded');
    }
    this.device = {
      realDevice, // TODO: temporarily, clean up T70688226
      // N.B. we model OS as string, not as enum, to make custom device types possible in the future
      os: realDevice.os,
      get isArchived() {
        return realDevice.isArchived;
      },
      get isConnected() {
        return realDevice.connected.get();
      },
      deviceType: realDevice.deviceType,

      onLogEntry(cb) {
        const handle = realDevice.addLogListener(cb);
        return () => {
          realDevice.removeLogListener(handle);
        };
      },
    };
  }

  protected initializePlugin(factory: () => any) {
    // To be called from constructory
    setCurrentPluginInstance(this);
    try {
      this.instanceApi = batched(factory)();
    } finally {
      // check if we have both an import handler and rootStates; probably dev error
      if (this.importHandler && Object.keys(this.rootStates).length > 0) {
        throw new Error(
          `A custom onImport handler was defined for plugin '${
            this.definition.id
          }', the 'persist' option of states ${Object.keys(
            this.rootStates,
          ).join(', ')} should not be set.`,
        );
      }
      if (this.initialStates) {
        if (this.importHandler) {
          try {
            batched(this.importHandler)(this.initialStates);
          } catch (e) {
            const msg = `Error occurred when importing date for plugin '${this.definition.id}': '${e}`;
            console.error(msg, e);
            message.error(msg);
          }
        } else {
          for (const key in this.rootStates) {
            if (key in this.initialStates) {
              this.rootStates[key].set(this.initialStates[key]);
            } else {
              console.warn(
                `Tried to initialize plugin with existing data, however data for "${key}" is missing. Was the export created with a different Flipper version?`,
              );
            }
          }
        }
      }
      this.initialStates = undefined;
      setCurrentPluginInstance(undefined);
    }
  }

  protected createBasePluginClient(): BasePluginClient {
    return {
      device: this.device,
      onActivate: (cb) => {
        this.events.on('activate', batched(cb));
      },
      onDeactivate: (cb) => {
        this.events.on('deactivate', batched(cb));
      },
      onDeepLink: (cb) => {
        this.events.on('deeplink', batched(cb));
      },
      onDestroy: (cb) => {
        this.events.on('destroy', batched(cb));
      },
      onExport: (cb) => {
        if (this.exportHandler) {
          throw new Error('onExport handler already set');
        }
        this.exportHandler = cb;
      },
      onImport: (cb) => {
        if (this.importHandler) {
          throw new Error('onImport handler already set');
        }
        this.importHandler = cb;
      },
      addMenuEntry: (...entries) => {
        for (const entry of entries) {
          const normalized = normalizeMenuEntry(entry);
          if (
            this.menuEntries.find(
              (existing) =>
                existing.label === normalized.label ||
                existing.action === normalized.action,
            )
          ) {
            throw new Error(`Duplicate menu entry: '${normalized.label}'`);
          }
          this.menuEntries.push(normalizeMenuEntry(entry));
        }
      },
      createPaste: this.flipperLib.createPaste,
      GK: this.flipperLib.GK,
    };
  }

  // the plugin is selected in the UI
  activate() {
    this.assertNotDestroyed();
    if (!this.activated) {
      this.activated = true;
      this.flipperLib.enableMenuEntries(this.menuEntries);
      this.events.emit('activate');
      this.flipperLib.logger.trackTimeSince(
        `activePlugin-${this.definition.id}`,
      );
    }
  }

  deactivate() {
    if (this.destroyed) {
      return;
    }
    if (this.activated) {
      this.activated = false;
      this.lastDeeplink = undefined;
      this.events.emit('deactivate');
    }
  }

  destroy() {
    this.assertNotDestroyed();
    this.deactivate();
    this.events.emit('destroy');
    this.destroyed = true;
  }

  triggerDeepLink(deepLink: unknown) {
    this.assertNotDestroyed();
    if (deepLink !== this.lastDeeplink) {
      this.lastDeeplink = deepLink;
      this.events.emit('deeplink', deepLink);
    }
  }

  exportStateSync() {
    // This method is mainly intended for unit testing
    if (this.exportHandler) {
      throw new Error(
        'Cannot export sync a plugin that does have an export handler',
      );
    }
    return Object.fromEntries(
      Object.entries(this.rootStates).map(([key, atom]) => [key, atom.get()]),
    );
  }

  async exportState(
    idler: Idler,
    onStatusMessage: (msg: string) => void,
  ): Promise<Record<string, any>> {
    if (this.exportHandler) {
      return await this.exportHandler(idler, onStatusMessage);
    }
    return this.exportStateSync();
  }

  isPersistable(): boolean {
    return !!this.exportHandler || Object.keys(this.rootStates).length > 0;
  }

  protected assertNotDestroyed() {
    if (this.destroyed) {
      throw new Error('Plugin has been destroyed already');
    }
  }

  abstract toJSON(): string;
}
