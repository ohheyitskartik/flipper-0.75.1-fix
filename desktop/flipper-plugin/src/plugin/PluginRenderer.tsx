/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import React, {memo, useEffect, createElement} from 'react';
import {SandyPluginContext} from './PluginContext';
import {SandyPluginInstance} from './Plugin';
import {SandyDevicePluginInstance} from './DevicePlugin';
import {BasePluginInstance} from './PluginBase';
import {TrackingScope} from '../ui/Tracked';

type Props = {
  plugin: SandyPluginInstance | SandyDevicePluginInstance;
};

/**
 * Component to render a Sandy plugin container
 */
export const SandyPluginRenderer = memo(({plugin}: Props) => {
  if (!plugin || !(plugin instanceof BasePluginInstance)) {
    throw new Error('Expected plugin, got ' + plugin);
  }
  useEffect(() => {
    plugin.activate();
    return () => {
      plugin.deactivate();
    };
  }, [plugin]);

  return (
    <TrackingScope scope={'plugin:' + plugin.definition.id}>
      <SandyPluginContext.Provider value={plugin}>
        {createElement(plugin.definition.module.Component)}
      </SandyPluginContext.Provider>
    </TrackingScope>
  );
});
