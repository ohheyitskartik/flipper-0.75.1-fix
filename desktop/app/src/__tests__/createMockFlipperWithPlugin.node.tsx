/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {createMockFlipperWithPlugin} from '../test-utils/createMockFlipperWithPlugin';
import {FlipperPlugin} from '../plugin';

interface PersistedState {
  count: 1;
}

class TestPlugin extends FlipperPlugin<any, any, any> {
  static id = 'TestPlugin';

  static defaultPersistedState = {
    count: 0,
  };

  static persistedStateReducer(
    persistedState: PersistedState,
    method: string,
    _payload: {},
  ) {
    if (method === 'inc') {
      return Object.assign({}, persistedState, {
        count: persistedState.count + 1,
      });
    }
    return persistedState;
  }

  render() {
    return null;
  }
}

test('can create a Fake flipper', async () => {
  const {
    client,
    device,
    store,
    sendMessage,
  } = await createMockFlipperWithPlugin(TestPlugin);
  expect(client).toBeTruthy();
  expect(device).toBeTruthy();
  expect(store).toBeTruthy();
  expect(sendMessage).toBeTruthy();
  expect(client.plugins.includes(TestPlugin.id)).toBe(true);
  expect(store.getState().connections).toMatchSnapshot();
  expect(store.getState().plugins).toMatchSnapshot();
  sendMessage('inc', {});
  expect(store.getState().pluginStates).toMatchInlineSnapshot(`
        Object {
          "TestApp#Android#MockAndroidDevice#serial#TestPlugin": Object {
            "count": 1,
          },
        }
      `);
});
