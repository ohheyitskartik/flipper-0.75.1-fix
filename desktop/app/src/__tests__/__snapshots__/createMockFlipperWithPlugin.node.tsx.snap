// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a Fake flipper 1`] = `
Object {
  "androidEmulators": Array [],
  "clients": Array [
    Object {
      "id": "TestApp#Android#MockAndroidDevice#serial",
      "query": Object {
        "app": "TestApp",
        "device": "MockAndroidDevice",
        "device_id": "serial",
        "os": "Android",
        "sdk_version": 4,
      },
    },
  ],
  "deepLinkPayload": null,
  "devices": Array [
    Object {
      "deviceType": "physical",
      "os": "Android",
      "serial": "serial",
      "title": "MockAndroidDevice",
    },
  ],
  "selectedApp": "TestApp#Android#MockAndroidDevice#serial",
  "selectedDevice": Object {
    "deviceType": "physical",
    "os": "Android",
    "serial": "serial",
    "title": "MockAndroidDevice",
  },
  "selectedPlugin": "TestPlugin",
  "staticView": null,
  "uninitializedClients": Array [],
  "userPreferredApp": "TestApp#Android#MockAndroidDevice#serial",
  "userPreferredDevice": "MockAndroidDevice",
  "userPreferredPlugin": "TestPlugin",
  "userStarredPlugins": Object {
    "TestApp": Array [
      "TestPlugin",
    ],
  },
}
`;

exports[`can create a Fake flipper 2`] = `
Object {
  "bundledPlugins": Map {},
  "clientPlugins": Map {
    "TestPlugin" => [Function],
  },
  "devicePlugins": Map {},
  "disabledPlugins": Array [],
  "failedPlugins": Array [],
  "gatekeepedPlugins": Array [],
  "loadedPlugins": Map {},
  "marketplacePlugins": Array [],
  "selectedPlugins": Array [],
}
`;
